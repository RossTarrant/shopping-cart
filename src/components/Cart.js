import { Link } from "react-router-dom";

export default function Cart({basket, resetBasket, changeQuantity, removeFromBasket}){

    const getTotal = () => {
        let total = 0;
        basket.forEach(item => {
            total += (item.quantity * item.price);
        });
        return total;
    }

    const onClickRemoveItem = (id) => {
        removeFromBasket(id);
    }

    const onClickChangeQuantity = (id, button) => {
        changeQuantity(id, button);
    }

    const onClickProceedToPayment = () => {
        alert(`Thank you for your purchase! You were charged £${getTotal()}.`)
        resetBasket();
    }

    const getBasket = () => {
        const itemBasket = basket.map((item) => 
            <div key={item.id} className={"border-black border-2 flex w-96 h-64"}>
                <img alt={"pic"} src={item.imgUrl} className={"w-32 h-32"}/>
                <div className={"flex flex-col items-center"}>
                    <h1 className={""}>{item.title}</h1>
                    <h2 className={""}>£{item.price}</h2>
                    <div className={"flex items-center gap-8"}>
                        <div className="flex">
                            <button onClick={e => onClickChangeQuantity(item.id, e.target.innerText)} className={"text-center"}>-</button>
                            <span className={"text-center w-16"} >{item.quantity}</span>
                            <button onClick={e => onClickChangeQuantity(item.id, e.target.innerText)} className={"text-center"}>+</button>
                        </div>
                        <button onClick={() => onClickRemoveItem(item.id)} className={"text-sm rounded-full bg-white-500 border-2 text-white border-red-600 bg-red-600 p-1 transition ease-out duration-300 hover:bg-red-800 hover:border-red-800"}>Delete</button>
                    </div> 
                </div>
            </div>
        )

        return itemBasket;
    }

    return(
        <div className={"flex flex-col items-center justify-center"}>
            <h2>Your Shopping Basket</h2>
            {basket.length === 0 ? <p>Your basket is empty! Add some items in the shop!</p> : null}
            {basket.length !== 0 ? <p>Your total is £{getTotal()}.00</p> : null}
            {basket.length !== 0 ? <p>Check the basket below and then click proceed button</p> : null}
            {basket.length !== 0 ? <Link to={'/'}><button onClick={onClickProceedToPayment} className={"border-black border-2 bg-slate-800 text-white rounded-full p-2"}>Proceed to payment</button></Link> : null}
            <div className={"flex flex-col gap-10 items-center font-jakarta"}>
                {getBasket()}
            </div>
        </div>
    )
}