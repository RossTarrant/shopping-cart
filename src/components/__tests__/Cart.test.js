import React from "react";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import Cart from "../Cart";

describe("Cart component", () => {

    const basket = [{
        id: 123,
        title: "Steve Madden",
        description: "A trainer from the astounding and incredible, Steve Madden!",
        price: 179,
        quantity: 2,
        imgUrl: "http://cdn.shopify.com/s/files/1/0083/3903/0082/products/POSS03S1_BLACKTAN_1_grande.jpg?v=1654077320",
    },
    {
        id: 456,
        title: "Match-E-Trainer",
        description: "It's pretty much just a brown shoe, nothing special!",
        price: 39,
        quantity: 1,
        imgUrl: "http://cdn.shopify.com/s/files/1/0083/3903/0082/products/SM19000020-04004-748_02_grande.jpg?v=1663106690",
    }]

    it("renders Cart", () => {
        const { container } = render(
            <BrowserRouter>
                <Cart basket={basket}/>
            </BrowserRouter>
        );
        expect(container).toMatchSnapshot();
    });

    it("calculates correct basket total", () => {
        render(
            <BrowserRouter>
                <Cart basket={basket}/>
            </BrowserRouter>
        );
        expect(screen.getByText("Your total is Â£397.00")).toBeTruthy();
    });

    it("updates basket total when deleting item from basket", () => {
        const onClickRemoveItem = jest.fn();
        render(
            <BrowserRouter>
                <Cart basket={basket} removeFromBasket={onClickRemoveItem}/>
            </BrowserRouter>
        );
        const button = screen.getAllByRole("button", { name: "Delete" })[0];
        userEvent.click(button);
        expect(onClickRemoveItem).toHaveBeenCalledTimes(1);
    });

});