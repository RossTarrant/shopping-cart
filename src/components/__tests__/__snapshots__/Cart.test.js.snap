// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component renders Cart 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center gap-4 text-center"
  >
    <h2
      class=" text-5xl p-4"
    >
      Your Shopping Basket
    </h2>
    <p>
      Your total is £
      397
      .00
    </p>
    <p>
      Check the basket below and then click proceed button
    </p>
    <a
      href="/"
    >
      <button
        class="bg-slate-800 text-white rounded-full p-2 hover:bg-slate-700"
      >
        Proceed to payment
      </button>
    </a>
    <div
      class="flex flex-col items-center font-jakarta"
    >
      <div
        class="grid grid-cols-2 p-4 m-4 justify-between border-2 border-neutral-900 rounded-lg shadow-md overflow-hidden"
      >
        <img
          alt="pic"
          class="w-60 h-60"
          src="http://cdn.shopify.com/s/files/1/0083/3903/0082/products/POSS03S1_BLACKTAN_1_grande.jpg?v=1654077320"
        />
        <div
          class="flex flex-col items-center justify-center gap-4"
        >
          <h1
            class="text-2xl"
          >
            Steve Madden
          </h1>
          <h2
            class=""
          >
            £
            179
          </h2>
          <div
            class="flex items-center gap-8"
          >
            <div
              class="flex"
            >
              <button
                class="text-center"
              >
                -
              </button>
              <span
                class="text-center w-16"
              >
                2
              </span>
              <button
                class="text-center"
              >
                +
              </button>
            </div>
            <button
              class="text-sm rounded-full bg-white-500 border-2 text-white border-red-600 bg-red-600 p-1 transition ease-out duration-300 hover:bg-red-800 hover:border-red-800"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
      <div
        class="grid grid-cols-2 p-4 m-4 justify-between border-2 border-neutral-900 rounded-lg shadow-md overflow-hidden"
      >
        <img
          alt="pic"
          class="w-60 h-60"
          src="http://cdn.shopify.com/s/files/1/0083/3903/0082/products/SM19000020-04004-748_02_grande.jpg?v=1663106690"
        />
        <div
          class="flex flex-col items-center justify-center gap-4"
        >
          <h1
            class="text-2xl"
          >
            Match-E-Trainer
          </h1>
          <h2
            class=""
          >
            £
            39
          </h2>
          <div
            class="flex items-center gap-8"
          >
            <div
              class="flex"
            >
              <button
                class="text-center"
              >
                -
              </button>
              <span
                class="text-center w-16"
              >
                1
              </span>
              <button
                class="text-center"
              >
                +
              </button>
            </div>
            <button
              class="text-sm rounded-full bg-white-500 border-2 text-white border-red-600 bg-red-600 p-1 transition ease-out duration-300 hover:bg-red-800 hover:border-red-800"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
